ts/session/utils/Attachments.ts

```diff
@@ -0,0 +1,164 @@
+import * as crypto from 'crypto';
+import { Attachment } from '../../types/Attachment';
+import { OpenGroup } from '../types';
+import {
+  AttachmentPointer,
+  Preview,
+  Quote,
+  QuotedAttachment,
+} from '../messages/outgoing';
+import { LokiAppDotNetServerInterface } from '../../../js/modules/loki_app_dot_net_api';
+
+interface UploadParams {
+  attachment: Attachment;
+  openGroup?: OpenGroup;
+  isAvatar?: boolean;
+  isRaw?: boolean;
+}
+
+interface RawPreview {
+  url?: string;
+  title?: string;
+  image: Attachment;
+}
+
+interface RawQuoteAttachment {
+  contentType?: string;
+  fileName?: string;
+  thumbnail?: Attachment;
+}
+
+interface RawQuote {
+  id?: number;
+  author?: string;
+  text?: string;
+  attachments?: Array<RawQuoteAttachment>;
+}
+
+// tslint:disable-next-line: no-unnecessary-class
+export class AttachmentUtils {
+  private constructor() {}
+
+  public static getDefaultServer(): LokiAppDotNetServerInterface {
+    return window.tokenlessFileServerAdnAPI;
+  }
+
+  public static async upload(params: UploadParams): Promise<AttachmentPointer> {
+    const { attachment, openGroup, isAvatar = false, isRaw = false } = params;
+    if (typeof attachment !== 'object' || attachment == null) {
+      throw new Error('Invalid attachment passed.');
+    }
+
+    if (!(attachment.data instanceof ArrayBuffer)) {
+      throw new TypeError(
+        `\`attachment.data\` must be an \`ArrayBuffer\`; got: ${typeof attachment.data}`
+      );
+    }
+
+    let server = this.getDefaultServer();
+    if (openGroup) {
+      const openGroupServer = await window.lokiPublicChatAPI.findOrCreateServer(
+        openGroup.server
+      );
+      if (!openGroupServer) {
+        throw new Error(
+          `Failed to get open group server: ${openGroup.server}.`
+        );
+      }
+      server = openGroupServer;
+    }
+
+    const pointer: AttachmentPointer = {
+      contentType: attachment.contentType
+        ? (attachment.contentType as string)
+        : undefined,
+      size: attachment.size,
+      fileName: attachment.fileName,
+      flags: attachment.flags,
+    };
+
+    let attachmentData: ArrayBuffer;
+
+    if (isRaw || openGroup) {
+      attachmentData = attachment.data;
+    } else {
+      server = this.getDefaultServer();
+      pointer.key = new Uint8Array(crypto.randomBytes(64));
+      const iv = new Uint8Array(crypto.randomBytes(16));
+      const data = await window.textsecure.crypto.encryptAttachment(
+        attachment.data,
+        pointer.key.buffer,
+        iv.buffer
+      );
+      pointer.digest = data.digest;
+      attachmentData = data.ciphertext;
+    }
+
+    const result = isAvatar
+      ? await server.putAvatar(attachmentData)
+      : await server.putAttachment(attachmentData);
```

Comment:
we can probably remove `makeAttachmentPointer` and some other methods from `sendmessage.js`

Reply:
I want to keep that in there until we remove the whole `sendmessage.js` file. The reason is that `makeAttachmentPointer` might be called from a weird case that we missed and we don't want sending to break because of it.